### PR TITLE
Introduce Filter/OutputSetting protocol

### PR DESCRIPTION
**Pull request by giginet**
# Motivation

I want to define custom Filter/Output setting. However, we can pass just only `FilterSetting` and `OutputSetting`.

# Description

I introduce `FilterSettingProtocol` and `OutputSettingProtocol`.
Thanks to them, we can define custom settings.

See the test code to know how to use. 
https://github.com/giginet/Puree-Swift/blob/32deb5e8a3a46378a75926bf9d8f5f26e91eb287/Tests/PureeTests/LoggerTests.swift#L87

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,4 +5,5 @@ before_script:
   - swiftlint
   - pod lib lint
 script:
+- swift test
```
Comment (giginet) [inline .travis.yml:8]: Run test via SPM

File: Sources/Puree/Logger.swift [hunk 2]
```diff
@@ -4,13 +4,13 @@ public final class Logger {
     public struct Configuration {
         public var logStore: LogStore
         public var dateProvider: DateProvider = DefaultDateProvider()
-        public var filterSettings: [FilterSetting] = []
-        public var outputSettings: [OutputSetting] = []
+        public var filterSettings: [FilterSettingProtocol] = []
+        public var outputSettings: [OutputSettingProtocol] = []
 
-        public init(logStore: LogStore,
+        public init(logStore: LogStore = FileLogStore.default,
```
Comment (giginet) [inline Sources/Puree/Logger.swift:10]: Use `FileLogStore` as default.

**Comment by giginet** [comment]
We can use these protocols to constrain required parameters.

```swift
struct CustomOutputSetting: OutputSettingProtocol {
    private let tableName: String

    init(tableName: String) {
        self.tableName = tableName
    }

    func makeOutput(_ logStore: LogStore) throws -> Output {
        return PVLogOutput(logStore: logStore, tagPattern: TagPattern(string: "pv2")!, options: ["table_name": tableName])
    }
}

CustomOutputSetting(tableName: "table_name")
```

File: Sources/Puree/Filter.swift [hunk 3]
```diff
@@ -2,13 +2,22 @@ import Foundation
 
 public typealias FilterOptions = [String: Any]
 
-public struct FilterSetting {
+public protocol FilterSettingProtocol {
+    func makeFilter() throws -> Filter
```
Comment (vincentisambart) [inline Sources/Puree/Filter.swift:6]: Wouldn't the following make things simpler and cleaner?
```swift
public protocol FilterSettingProtocol {
    var filterMaker: () -> Filter { get }
}
```

File: Sources/Puree/Output.swift [hunk 4]
```diff
@@ -2,13 +2,22 @@ import Foundation
 
 public typealias OutputOptions = [String: Any]
 
-public struct OutputSetting {
+public protocol OutputSettingProtocol {
+    func makeOutput(_ logStore: LogStore) throws -> Output
```
Comment (vincentisambart) [inline Sources/Puree/Output.swift:6]: Similar to `FilterSettingProtocol`, wouldn't the following make things simpler and clearer?
```swift
public protocol OutputSettingProtocol {
    var outputMaker: (_ logStore: LogStore) throws -> Output { get }
}
```

File: Tests/PureeTests/LoggerTests.swift [hunk 5]
```diff
@@ -84,6 +84,53 @@ class LoggerTests: XCTestCase {
...
+                return PVLogOutput(logStore: logStore, tagPattern: TagPattern(string: "pv2")!, options: ["table_name": tableName])
+            }
+        }
+
+        let configuration = Logger.Configuration(logStore: logStore,
+                                                 dateProvider: DefaultDateProvider(),
+                                                 filterSettings: [
+                                                    FilterSetting(PVLogFilter.self, tagPattern: TagPattern(string: "pv")!),
+                                                    CustomFilterSetting(tableName: "pv_log"),
+                                                    FilterSetting(PVLogFilter.self, tagPattern: TagPattern(string: "pv.*")!)
```
Comment (vincentisambart) [inline Tests/PureeTests/LoggerTests.swift]: [nits] `,` at the end of the line.

Comment (giginet) [reply Tests/PureeTests/LoggerTests.swift]: https://github.com/cookpad/Puree-Swift/pull/1/commits/b261a42a26a8a231c0b4af5ff498458fa56b9cf4

I added `,` and enable `mandatory_comma`.

File: Sources/Puree/Filter.swift (hunk 3 above)
Comment (vincentisambart) [reply Sources/Puree/Filter.swift:6]: After thinking more about it, it seems to me that `FilterSetting` is not even useful. Directly passing a `Filter` to `Configuration` would achieve the same result...

File: Sources/Puree/Output.swift (hunk 4 above)
Comment (vincentisambart) [reply Sources/Puree/Output.swift:6]: In fact your `OutputSetting` could be just a `(_ logStore: LogStore) -> Output`...

File: Sources/Puree/Filter.swift (hunk 3 above)
Comment (giginet) [reply Sources/Puree/Filter.swift:6]: I agree. I'll modify this interface next major version.

**Review by vincentisambart — APPROVED** [review]
If I was maintaining Puree-Swift I wouldn't wait before improving the interface but well...

**Pull request merged**